# 블랙잭
## 문제
카지노에서 제일 인기 있는 게임 블랙잭의 규칙은 상당히 쉽다. 카드의 합이 21을 넘지 않는 한도 내에서, 카드의 합을 최대한 크게 만드는 게임이다. 블랙잭은 카지노마다 다양한 규정이 있다.

한국 최고의 블랙잭 고수 김정인은 새로운 블랙잭 규칙을 만들어 상근, 창영이와 게임하려고 한다.

김정인 버전의 블랙잭에서 각 카드에는 양의 정수가 쓰여 있다. 그 다음, 딜러는 N장의 카드를 모두 숫자가 보이도록 바닥에 놓는다. 그런 후에 딜러는 숫자 M을 크게 외친다.

이제 플레이어는 제한된 시간 안에 N장의 카드 중에서 3장의 카드를 골라야 한다. 블랙잭 변형 게임이기 때문에, 플레이어가 고른 카드의 합은 M을 넘지 않으면서 M과 최대한 가깝게 만들어야 한다.

N장의 카드에 써져 있는 숫자가 주어졌을 때, M을 넘지 않으면서 M에 최대한 가까운 카드 3장의 합을 구해 출력하시오.

## 입력
첫째 줄에 카드의 개수 N(3 ≤ N ≤ 100)과 M(10 ≤ M ≤ 300,000)이 주어진다. 둘째 줄에는 카드에 쓰여 있는 수가 주어지며, 이 값은 100,000을 넘지 않는 양의 정수이다.

합이 M을 넘지 않는 카드 3장을 찾을 수 있는 경우만 입력으로 주어진다.

## 출력
첫째 줄에 M을 넘지 않으면서 M에 최대한 가까운 카드 3장의 합을 출력한다.

In [ ]:
N, M = list(map(int,input().split()))
elements = list(map(int,input().split()))

best_partial_sum = 0
for i in elements:
    elements2 = elements.copy()
    elements2.remove(i)
    for j in elements2:
        elements3 = elements2.copy()
        elements3.remove(j)
        for k in elements3:
            partial_sum = i+j+k
            if partial_sum > M:
                continue
            elif partial_sum > best_partial_sum:
                best_partial_sum = partial_sum
print(best_partial_sum)

10 500
93 181 245 214 315 36 185 138 216 295
497


# 분배합
## 문제
어떤 자연수 N이 있을 때, 그 자연수 N의 분해합은 N과 N을 이루는 각 자리수의 합을 의미한다. 어떤 자연수 M의 분해합이 N인 경우, M을 N의 생성자라 한다. 예를 들어, 245의 분해합은 256(=245+2+4+5)이 된다. 따라서 245는 256의 생성자가 된다. 물론, 어떤 자연수의 경우에는 생성자가 없을 수도 있다. 반대로, 생성자가 여러 개인 자연수도 있을 수 있다.

자연수 N이 주어졌을 때, N의 가장 작은 생성자를 구해내는 프로그램을 작성하시오.

## 입력
첫째 줄에 자연수 N(1 ≤ N ≤ 1,000,000)이 주어진다.

## 출력
첫째 줄에 답을 출력한다. 생성자가 없는 경우에는 0을 출력한다.

In [ ]:
N = int(input())
no_generator = False
for n in range(N):
    if n + sum(list(map(int,list(str(n)))))==N:
        print(n)
        no_generator = True
        break
if no_generator == False:
    print(0)

256
245


# 덩치
## 문제
우리는 사람의 덩치를 키와 몸무게, 이 두 개의 값으로 표현하여 그 등수를 매겨보려고 한다. 어떤 사람의 몸무게가 x kg이고 키가 y cm라면 이 사람의 덩치는 (x, y)로 표시된다. 두 사람 A 와 B의 덩치가 각각 (x, y), (p, q)라고 할 때 x > p 그리고 y > q 이라면 우리는 A의 덩치가 B의 덩치보다 "더 크다"고 말한다. 예를 들어 어떤 A, B 두 사람의 덩치가 각각 (56, 177), (45, 165) 라고 한다면 A의 덩치가 B보다 큰 셈이 된다. 그런데 서로 다른 덩치끼리 크기를 정할 수 없는 경우도 있다. 예를 들어 두 사람 C와 D의 덩치가 각각 (45, 181), (55, 173)이라면 몸무게는 D가 C보다 더 무겁고, 키는 C가 더 크므로, "덩치"로만 볼 때 C와 D는 누구도 상대방보다 더 크다고 말할 수 없다.

N명의 집단에서 각 사람의 덩치 등수는 자신보다 더 "큰 덩치"의 사람의 수로 정해진다. 만일 자신보다 더 큰 덩치의 사람이 k명이라면 그 사람의 덩치 등수는 k+1이 된다. 이렇게 등수를 결정하면 같은 덩치 등수를 가진 사람은 여러 명도 가능하다. 아래는 5명으로 이루어진 집단에서 각 사람의 덩치와 그 등수가 표시된 표이다.

| 이름 | (몸무게, 키) | 덩치 등수 |
|-----|-----|-----|
| A | (55, 185) | 2 |
| B | (58, 183) | 2 |
| C | (88, 186) | 1 |
| D | (60, 175) | 2 |
| E | (46, 155) | 5 |

위 표에서 C보다 더 큰 덩치의 사람이 없으므로 C는 1등이 된다. 그리고 A, B, D 각각의 덩치보다 큰 사람은 C뿐이므로 이들은 모두 2등이 된다. 그리고 E보다 큰 덩치는 A, B, C, D 이렇게 4명이므로 E의 덩치는 5등이 된다. 위 경우에 3등과 4등은 존재하지 않는다. 여러분은 학생 N명의 몸무게와 키가 담긴 입력을 읽어서 각 사람의 덩치 등수를 계산하여 출력해야 한다.

## 입력
첫 줄에는 전체 사람의 수 N이 주어진다. 그리고 이어지는 N개의 줄에는 각 사람의 몸무게와 키를 나타내는 양의 정수 x와 y가 하나의 공백을 두고 각각 나타난다.

## 출력
여러분은 입력에 나열된 사람의 덩치 등수를 구해서 그 순서대로 첫 줄에 출력해야 한다. 단, 각 덩치 등수는 공백문자로 분리되어야 한다.

## 제한
2 ≤ N ≤ 50
10 ≤ x, y ≤ 200


In [ ]:
N = int(input())
xs = []
ys = []
for n in range(N):
    xy = input()
    x = int(xy.split()[0])
    y = int(xy.split()[1])
    xs.append(x)
    ys.append(y)

xys = list(zip(xs,ys))

for i1,i2 in xys:
    temp_xys = xys.copy()
    temp_xys.remove((i1,i2))
    cnt = 1
    for j1,j2 in temp_xys:
        if (i1<j1) & (i2<j2):
            cnt+=1
    print(cnt,end=' ')

5
55 185
58 183
88 176
60 175
46 155
2 2 1 2 5 

# 체스판 다시 칠하기
## 문제
지민이는 자신의 저택에서 MN개의 단위 정사각형으로 나누어져 있는 M×N 크기의 보드를 찾았다. 어떤 정사각형은 검은색으로 칠해져 있고, 나머지는 흰색으로 칠해져 있다. 지민이는 이 보드를 잘라서 8×8 크기의 체스판으로 만들려고 한다.

체스판은 검은색과 흰색이 번갈아서 칠해져 있어야 한다. 구체적으로, 각 칸이 검은색과 흰색 중 하나로 색칠되어 있고, 변을 공유하는 두 개의 사각형은 다른 색으로 칠해져 있어야 한다. 따라서 이 정의를 따르면 체스판을 색칠하는 경우는 두 가지뿐이다. 하나는 맨 왼쪽 위 칸이 흰색인 경우, 하나는 검은색인 경우이다.

보드가 체스판처럼 칠해져 있다는 보장이 없어서, 지민이는 8×8 크기의 체스판으로 잘라낸 후에 몇 개의 정사각형을 다시 칠해야겠다고 생각했다. 당연히 8*8 크기는 아무데서나 골라도 된다. 지민이가 다시 칠해야 하는 정사각형의 최소 개수를 구하는 프로그램을 작성하시오.

## 입력
첫째 줄에 N과 M이 주어진다. N과 M은 8보다 크거나 같고, 50보다 작거나 같은 자연수이다. 둘째 줄부터 N개의 줄에는 보드의 각 행의 상태가 주어진다. B는 검은색이며, W는 흰색이다.

## 출력
첫째 줄에 지민이가 다시 칠해야 하는 정사각형 개수의 최솟값을 출력한다.

In [89]:
def inequal_str(xx,yy):
    result = 0
    for x,y in zip(xx,yy):
        if x!=y:
            result +=1
    return result
    


N,M = tuple(map(int,input().split()))
ii = N-8+1
jj = M-8+1

board = []
for _ in range(N):
    board.append(input())
    
repair_cnt_list = []
for i in range(ii):
    for j in range(jj):
        repair_cnt_line1 = 0
        rapair_cnt_line2 = 0
        for line_num,line in enumerate(board[i:i+8]):
            board_line = line[j:j+8]
            repair_cnt_lineBW = inequal_str(board_line,'BWBWBWBW')
            repair_cnt_lineWB = inequal_str(board_line,'WBWBWBWB')
            if line_num % 2==0:
                repair_cnt_line1 += repair_cnt_lineBW
                rapair_cnt_line2 += repair_cnt_lineWB
            else:
                repair_cnt_line1 += repair_cnt_lineWB
                rapair_cnt_line2 += repair_cnt_lineBW
        repair_cnt = min([repair_cnt_line1, rapair_cnt_line2])
        repair_cnt_list.append(repair_cnt)
print(min(repair_cnt_list))

9 23
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBW
31


# 영화감독 숌
## 문제
666은 종말을 나타내는 숫자라고 한다. 따라서, 많은 블록버스터 영화에서는 666이 들어간 제목을 많이 사용한다. 영화감독 숌은 세상의 종말 이라는 시리즈 영화의 감독이다. 조지 루카스는 스타워즈를 만들 때, 스타워즈 1, 스타워즈 2, 스타워즈 3, 스타워즈 4, 스타워즈 5, 스타워즈 6과 같이 이름을 지었고, 피터 잭슨은 반지의 제왕을 만들 때, 반지의 제왕 1, 반지의 제왕 2, 반지의 제왕 3과 같이 영화 제목을 지었다.

하지만 숌은 자신이 조지 루카스와 피터 잭슨을 뛰어넘는다는 것을 보여주기 위해서 영화 제목을 좀 다르게 만들기로 했다.

종말의 숫자란 어떤 수에 6이 적어도 3개이상 연속으로 들어가는 수를 말한다. 제일 작은 종말의 숫자는 666이고, 그 다음으로 큰 수는 1666, 2666, 3666, .... 과 같다.

따라서, 숌은 첫 번째 영화의 제목은 세상의 종말 666, 두 번째 영화의 제목은 세상의 종말 1666 이렇게 이름을 지을 것이다. 일반화해서 생각하면, N번째 영화의 제목은 세상의 종말 (N번째로 작은 종말의 숫자) 와 같다.

숌이 만든 N번째 영화의 제목에 들어간 숫자를 출력하는 프로그램을 작성하시오. 숌은 이 시리즈를 항상 차례대로 만들고, 다른 영화는 만들지 않는다.

## 입력
첫째 줄에 숫자 N이 주어진다. N은 10,000보다 작거나 같은 자연수이다.

## 출력
첫째 줄에 N번째 영화의 제목에 들어간 수를 출력한다.

In [ ]:
# 해당 셀은 틀림
decs = list(map(str,range(0,1000)))
decs_back = list(map(str,range(10))) + list(map(lambda x:str(x).zfill(2),range(100))) + list(map(lambda x:str(x).zfill(3),range(1000)))+['']
series_num = []
for i in decs:
    for j in decs_back:
        series_num.append(int(i+'666'+j))
series_sorted = sorted(list(set(series_num)))

N = int(input())
print(series_sorted[N-1])

10000
3366669


In [ ]:
# 3366669는 위 셀의 결과로 얻은 휴리스틱한 값
series_num = []
cond = len(series_num) < 10000
num = 666
# while cond:
for i in range(3366669):
    if '666' in str(i):
        series_num.append(i)
    if len(series_num) > 10000:
        break
series_sorted = sorted(series_num)

N = int(input())
print(series_sorted[N-1])

500
166699
